tests/integration/targets/vmware_guest_storage_policy/tasks/main.yml

```diff
@@ -0,0 +1,214 @@
+      vmware_category:
+        hostname: "{{ vcenter_hostname }}"
+        username: "{{ vcenter_username }}"
+        password: "{{ vcenter_password }}"
+        validate_certs: false
+        category_name: "{{ category }}"
+        category_cardinality: "multiple"
+        state: "present"
+      register: category_create
+
+    - name: "Setup: Set category id"
+      set_fact:
+        category_id: "{{ category_create['category_results']['category_id'] }}"
+
+    - name: "Setup: Create tag"
+      vmware_tag:
+        hostname: "{{ vcenter_hostname }}"
+        username: "{{ vcenter_username }}"
+        password: "{{ vcenter_password }}"
+        validate_certs: false
+        tag_name: "{{ tag }}"
+        category_id: "{{ category_id }}"
+        state: "present"
+      register: tag_create
+
+    - name: "Setup: Create vSphere tag-based storage policy: {{ storage_policy }}"
+      vmware_vm_storage_policy:
+        hostname: "{{ vcenter_hostname }}"
+        username: "{{ vcenter_username }}"
+        password: "{{ vcenter_password }}"
+        validate_certs: false
+        name: "{{ storage_policy }}"
+        description: "test storage policy for vmware_guest_storage_policy"
+        tag_category: "{{ category }}"
+        tag_name: "{{ tag }}"
+        tag_affinity: true
+        state: "present"
+      register: new_sp
+
+    - name: "001: Assign storage policy to VM check_mode test"
+      vmware_guest_storage_policy: &vm_home_args
+        hostname: "{{ vcenter_hostname }}"
+        username: "{{ vcenter_username }}"
+        password: "{{ vcenter_password }}"
+        datacenter: "{{ dc1 }}"
+        validate_certs: false
+        name: "{{ virtual_machines[0].name }}"
+        vm_home: "{{ storage_policy }}"
+      register: vm_home_sp_cm
+      check_mode: true
+
+    - name: "001: assert that changes were made"
+      assert:
+        that:
+          - vm_home_sp_cm is changed
+
+    - name: "001: assert that changed_policies.vm_home is properly set"
+      assert:
+        that:
+          - vm_home_sp_cm.changed_policies.vm_home == storage_policy
+
+    - name: "002: Assign storage policy to VM"
+      vmware_guest_storage_policy:
+        <<: *vm_home_args
+      register: vm_home_sp
+
+    - name: "002: assert that changes were made"
+      assert:
+        that:
+          - vm_home_sp is changed
+
+    - name: "002: assert that changed_policies.vm_home is properly set"
+      assert:
+        that:
+          - vm_home_sp.changed_policies.vm_home == storage_policy
+
+
+    - name: "003: Assign storage policy to VM idempotence test"
+      vmware_guest_storage_policy:
+        <<: *vm_home_args
+      register: vm_home_sp_idp
+
+    - name: "003: assert that changes were not made"
+      assert:
+        that:
+          - vm_home_sp_idp is not changed
+
+    - name: "003: assert that changed_policies.vm_home is not set"
+      assert:
+        that:
+          - vm_home_sp.changed_policies.vm_home == ""
```

Comment:
```suggestion
          - vm_home_sp.changed_policies.vm_home == storage_policy
```

Reply:
@sky-joker for this idempotence test vm_home should not be set, is it really should be this empty string. Is that not happening?

Reply:
Yes, it is set.
The following is the result.

```yaml
(snip)
    - debug: var=vm_home_sp.changed_policies.vm_home

    - name: "003: assert that changed_policies.vm_home is not set"
      assert:
        that:
          - vm_home_sp.changed_policies.vm_home == ""
(snip)
```

```
(snip)
TASK [debug] *************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "vm_home_sp.changed_policies.vm_home": "vm_guest_storage_profile_policy001"
}

TASK [003: assert that changed_policies.vm_home is not set] **************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {
    "assertion": "vm_home_sp.changed_policies.vm_home == \"\"",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
(snip)
```

Reply:
@sky-joker it's just evaluating the wrong register, everything else is right.
Could you change it from vm_home_sp to vm_home_sp_idp on line 111?

Reply:
ok :)